ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/upnp/PantheonUpnpServiceConfiguration.java

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.p2p.upnp;
+
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import org.jupnp.DefaultUpnpServiceConfiguration;
+import org.jupnp.UpnpServiceConfiguration;
+import org.jupnp.binding.xml.DeviceDescriptorBinder;
+import org.jupnp.binding.xml.ServiceDescriptorBinder;
+import org.jupnp.binding.xml.UDA10DeviceDescriptorBinderImpl;
+import org.jupnp.binding.xml.UDA10ServiceDescriptorBinderImpl;
+import org.jupnp.model.Namespace;
+import org.jupnp.model.message.UpnpHeaders;
+import org.jupnp.model.meta.RemoteDeviceIdentity;
+import org.jupnp.model.meta.RemoteService;
+import org.jupnp.model.types.ServiceType;
+import org.jupnp.transport.impl.DatagramIOConfigurationImpl;
+import org.jupnp.transport.impl.DatagramIOImpl;
+import org.jupnp.transport.impl.DatagramProcessorImpl;
+import org.jupnp.transport.impl.GENAEventProcessorImpl;
+import org.jupnp.transport.impl.MulticastReceiverConfigurationImpl;
+import org.jupnp.transport.impl.MulticastReceiverImpl;
+import org.jupnp.transport.impl.NetworkAddressFactoryImpl;
+import org.jupnp.transport.impl.SOAPActionProcessorImpl;
+import org.jupnp.transport.impl.jetty.StreamClientConfigurationImpl;
+import org.jupnp.transport.spi.DatagramIO;
+import org.jupnp.transport.spi.DatagramProcessor;
+import org.jupnp.transport.spi.GENAEventProcessor;
+import org.jupnp.transport.spi.MulticastReceiver;
+import org.jupnp.transport.spi.NetworkAddressFactory;
+import org.jupnp.transport.spi.SOAPActionProcessor;
+import org.jupnp.transport.spi.StreamClient;
+import org.jupnp.transport.spi.StreamServer;
+
+class PantheonUpnpServiceConfiguration implements UpnpServiceConfiguration {
```

Comment:
This code grew out of a series of ugly workarounds, and was reorganized / improved upon by @shemnon . I'm guessing that the choice to implement `UpnpServiceConfiguration` was explicit, but I'll let him chime in if he wants.

Reply:
We needed to cut out the stream server and the other bucket of threads that would be created for that.  Their executor also had a bug that hung upnp discovery and we are able to bring the fix in rather than wait for the upstream library to publish a fix.